roqoqo/src/operations/mod.rs

```diff
@@ -882,3 +881,191 @@ pub(crate) fn check_valid_mapping(mapping: &HashMap<usize, usize>) -> Result<(),
+    /// Returns `mode_1` bosonic mode of two bosonic mode Operation.
+    fn mode_1(&self) -> &usize;
+}
+
+/// Trait for operations acting on multiple (more than two) bosonic modes.
+///
+/// # Example
+/// ```
+/// use roqoqo::operations::{InvolveModes};
+/// ```
+///
+pub trait OperateMultiMode:
+    Operate + InvolveModes + SubstituteModes + Clone + PartialEq + SupportedVersion
+{
+    /// Returns vector of bosonic modes operation is acting on in descending order of significance
+    fn modes(&self) -> &Vec<usize>;
+}
+
+/// Trait for unitary operations acting on exactly one bosonic mode.
+///
+/// Implements the general single bosonic mode unitary gates that can be brought into the form:
+///
+/// U =exp(i * φ) * [[Re(α)+i * Im(α), -Re(β) + i * Im(β)], [Re(β) + i * Im(β) , Re(α) - i * Im(α) ]].
+///
+/// These gates can be parametrized by five real parameters:
+///
+/// * `alpha_r` - The real part Re(α) of the on-diagonal elements of the single-mode unitary.
+/// * `alpha_i` - The imaginary part Im(α) of the on-diagonal elements of the single-mode unitary.
+/// * `beta_r` - The real part Re(β) of the off-diagonal elements of the single-mode unitary.
+/// * `beta_i` - The imaginary part Im(β) of the off-diagonal elements of the single-mode unitary.
+/// * `global_phase` - The global phase φ of the single-mode unitary.
+///
+/// These are the single bosonic mode gates that are performed in the Circuit(), and are then translated
+/// to quantum hardware through the relevant backend. Two-mode gates are also available.
+///
+/// # Example
+/// ```
+/// use qoqo_calculator::CalculatorFloat;
+/// use roqoqo::operations::{OperateSingleModeGate};
+/// use std::f64::consts::PI;
+/// ```
+///
+pub trait OperateSingleModeGate:
+    Operate
+    + OperateModeGate
+    + InvolveModes
+    + SubstituteModes
+    + OperateSingleMode
+    + Clone
+    + PartialEq
+    + SupportedVersion
+    + std::fmt::Debug
+{
+}
+
+/// Trait for all Operations operating on or affecting exactly two bosonic modes.
+///
+/// # Example
+/// ```
+/// use roqoqo::operations::{OperateTwoModeGate};
+/// use qoqo_calculator::CalculatorFloat;
+/// ```
+///
+pub trait OperateTwoModeGate:
+    Operate
+    + OperateModeGate
+    + OperateTwoMode
+    + InvolveModes
+    + SubstituteModes
+    + Clone
+    + PartialEq
+    + SupportedVersion
+{
+}
+
+/// Trait for all Operations operating on or affecting more than two bosonic modes.
+///
+/// # Example
+/// ```
+/// use roqoqo::operations::{OperateMultiModeGate};
+/// use roqoqo::Circuit;
+/// use qoqo_calculator::CalculatorFloat;
+///
+/// let mut circuit = Circuit::new();
+/// ```
+///
+pub trait OperateMultiModeGate:
+    Operate
+    + OperateModeGate
+    + OperateMultiMode
+    + InvolveModes
+    + SubstituteModes
+    + Clone
+    + PartialEq
+    + SupportedVersion
+{
+    /// Returns a decomposition of the multi-mode operation using a circuit with two-mode-operations.
+    fn circuit(&self) -> crate::Circuit;
```

Comment:
I don't think we want to require a circuit decomposition at the start. We could also simply defer multi mode support till later so we do not commit here for this release (put it behind an unstable feature gate)